Wip rd (#1)
learn-co-curriculum/js-advanced-functions-context-lab: README.md

intro(asgardianBrothers[0], phrase) === introWithContext.apply(asgardianBrothers[0], [phrase]) //=> true

let huffyPhrase = "I was falling for thirty minutes!"
intro(asgardianBrothers[1], huffyPhrase) === introWithContext.call(asgardianBrothers[1], huffyPhrase) //=> true
intro(asgardianBrothers[1], huffyPhrase) === introWithContext.apply(asgardianBrothers[1], [huffyPhrase]) //=> true
```

In the previous lesson, we wrote functions like `intro`. They took the record
as an argument. In fact, if we look at the `solution` for the previous lesson
we'll see that multiple functions have the same, first parameter: `employee`,
the record.

```js
let createTimeInEvent = function(employee, dateStamp){ }
let createTimeOutEvent = function(employee, dateStamp){ }
let hoursWorkedOnDate = function(employee, soughtDate){ }
```

What if we told JavaScript that instead of the record being parameter (in
addition to a catchphrase), it could be assumed as a _context_ and thus
accessible via `this`.

That's exactly what we do with `introWithContext`. The `introWithContext` has
no expected parameters (except for a catchphrase). Both `call` and `apply` have

> take a `thisArg` argument as their first argument: that argument becomes the

Author: typo

`this` _inside_ the function. In the case of `call`, anything after the
`thisArg` gets passed to the function like arguments. In the case of `apply`,
the contents in the `Array` get destructured and passed to the function like
arguments.

> **ES6 ALERT**: Some might wonder, if we have destructuring of Arrays, why do
> we need both `call` _and_ `apply`. Destructuring is a relatively new arrival

> > to JavaScript, so had two separate methods.

Author: fix

# JavaScript Advanced Functions: Context

## Learning Goals

* Define execution context
* Define `this`
* Access implicitly-set global object in a JavaScript engine
* Access implicitly-set global object in a function call
* Prevent implicitly setting in function calls with `use strict`
* Access implicitly-set parent object in a contained function expression
* Access implicitly-set new object in object-oriented programming constructor
* Explicitly override context with `call` and `apply`
* Explicitly lock context for a function with `bind`

## Introduction

In the previous lesson we provided definitions of:

* "Execution Context"
* `this`
* `call`
* `apply`
* `bind`

but then built an application that didn't use them. However, with a
record-oriented application built, we will have a shared context to understand
these challenging concepts in JavaScript (there's a pun in there, but you won't
get it until the end of this lesson).

## Define Execution Context

When a function in JavaScript ***is called***, it is provided an _execution
context_.

The _execution context_ is a JavaScript `Object` that is either implicitly or

> explicitly passed along with the function at the time of the function's call.

Reviewer: ```suggestion
explicitly passed at the time of the function's call.
```
consider striking

The implicit way of passing a context with a function is something we have to
memorize and accept as part of the nature of JavaScript.

The tools for explicitly passing a context at function call-time are the
methods `call`, `apply`, and `bind.`

## Define `this`

The JavaScript keyword `this` returns the current _execution context_. Whether
that context was passed explicitly or implicitly, `this` returns it.

> 

Reviewer: I think we could see an example where we log `this` from a simple function, and indicate that _what is logged will change based on how we call the function_. That's the big picture point, so seems important to see code early that reinforces it

We'll now identify places where the _execution context_ is implicitly set.

### Access Implicitly-Set Global Object in a JavaScript Engine

```javascript
this //=> Window
typeof this //=> "object"
this === window //=> true
```

In a brand new Chrome console, try the code above. All the code you write in a
JavaScript file can be thought of as "one big function" whose context is set to
the `Window` object. Thus `this === window` is `true`.

We can try the same thing out in NodeJS. If you have `node` installed on your
computer, open the Node REPL by typing `node`.

```js
this === global //=> true
```

Node calls the default _execution context_ `global`, but otherwise it's a
global object like `window`.

> **STICKY PROBLEM**: Keen-eyed readers might be thinking: "But is `this ===
> Window` also `true`?" It is not. The `window` property is part of the
> `Window` global object; yes, the `Window` `Object` has a property that points

> > to itself.

Reviewer: I'm not sure this is the right explanation for this - `Window` is the type of `window`, but it doesn't have a property `window` (at least in Chrome 74) 

```js
Window.window
// undefined
```

I think we could mention that the type of `window` is `Window`, if we've been used to seeing types of things in the console. Otherwise, I think this is pretty confusing.

Reviewer: If we can avoid getting into an aside on the type of the global object, I think it might be more of a distraction than the point

Reviewer: Reading https://developer.mozilla.org/en-US/docs/Web/API/Window/window and seeing the confusion - `window.window` is defined, but `Window.window` is not. It's an instance-level accessor, not a class-level accessor, in OO terms.

>
> Due to the strictness of `===`, a property pointing to a thing is not the
> same thing as the thing itself. As such `this === Window` is `false`, but `this
> === window` is `true` _despite the fact_ that they both ultimately point to
> the same thing.
>
> If all this strikes you as a bit fussy and uninteresting, it's not essential
> understanding for the rest of this module. For more on this consult
> [Window.window][w1].

### Access Implicitly-Set Global Object in a Function Call

***All function declarations and function expressions*** you have called to
this point have also had an _execution context_, you just didn't know to ask
for it with `this`.

```js
let contextReturner = function() {
  return this
}

contextReturner() //=> window
contextReturner() === window //=> true
```

When you invoked a function, the global context was _implicitly_ passed.

JavaScript does this so that by default our functions can access any properties

> on that most-global object, `window` (or `global` in Node). Consider:

Reviewer: Is that the reasoning? I assumed that they decided that functions should have a context, and chose the global object as a default because they made a mistake. You can always access `window` or `global` from inside a function, so making it _also_ accessible as `this` doesn't seem like it helps with making it more accessible

Author: silent dot here on global obj

Author: tie to speak ex

```js
let locationReturner = function() {
  return this.location.host
}

locationReturner() //=> URL host serving this page e.g. developer.mozilla.org
```

Doesn't that seem _nice_ and handy to do? In an effort to be nice to us
programmers, JavaScript's implementers made the global object available inside
all functions.

> > **NEUROLINGUISTIC/ANTHROPOLOGICAL ASIDE**: Some people think that `this` is a

Reviewer: I like this note, but it seems like it breaks the flow. Consider moving to the end of the section?

function's default context is still the global object:

```js

function a() {
  return function b() {
    return this;
  }
}

a()() === window //=> true

### Prevent Implicitly Setting in Function Calls With `use strict`

We wish we could say that the default context was **always** the global object.
It'd make things simple.

However, in JavaScript, if the engine sees the `String` "use strict" inside a
function, it will _stop_ passing the implicit _execution context_ of the global
object.  If JavaScript sees `"use strict"` at the top of a JavaScript code
file, it will apply this rule (and other strict behaviors) to _all functions_.

```js
function looseyGoosey() {
  return this
}

function noInferringAllowed() {
  "use strict"
  return this
}

looseyGoosey() === window; //=> true
noInferringAllowed() === undefined //=> true
```

There are really no guidelines as to which you'll see more. Some programmers
think `strict` stops confusing bugs (seems wise!); others think it's an obvious
rule of the language and squelching it is against the language's love of
functions (a decent argument!). Generally, we advise you to think of the

> "default mode" as the one that permits an _implicit_ presumption of context.

Reviewer: About as clear and friendly a paragraph on default vs. strict mode as I've seen.

For more on strict-mode, see the Resources.

### Access Implicitly-Set Parent Object in a Contained Function Expression

This (no pun intended) is simple, thankfully: functions inside Objects have

> `this` pointing to their container so that they can access other sibling

Reviewer: Not exactly how this rule works - it's the 'thing to the left of the dot' that is passed as the context. Hence
```js
byronPoodle.warn()
// LOG: Byron issues an ear-rupturing atomic bark when he hears noises in the apartment hallway
w = byronPoodle.warn
w()
// issues an undefined when he hears undefined
```

So, this is misleading at best

intro(asgardianBrothers[0], phrase) === introWithContext.apply(asgardianBrothers[0], [phrase]) //=> true

let complaint = "I was falling for thirty minutes!"
intro(asgardianBrothers[1], complaint) === introWithContext.call(asgardianBrothers[1], complaint) //=> true
intro(asgardianBrothers[1], complaint) === introWithContext.apply(asgardianBrothers[1], [complaint]) //=> true
```

In the previous lesson, we wrote functions in the style of `intro`. They took
the record as an argument. In fact, if we look at the `solution` for the
previous lesson we'll see that multiple functions have the same, first
parameter: `employee`, the record.

```js
let createTimeInEvent = function(employee, dateStamp){ /* */ }
let createTimeOutEvent = function(employee, dateStamp){ /* */ }
let hoursWorkedOnDate = function(employee, soughtDate){ /* */ }
```

What if we told JavaScript that instead of the record being a _parameter_ (in
addition to a phrase), it could be assumed as a _context_ and thus accessible
via `this`. That's what we're doing with the function `introWithContext` as
invoked with either `call` or `apply`.

> The `introWithContext` function has expects only a catchphrase. Both `call` and

Reviewer: ```suggestion
The `introWithContext` function expects only a catchphrase as an argument. Both `call` and
```

the last few lessons.

Put your code in `index.js`.

## A Mystery on the Horizon

You'll notice that in this lab we give you the implementation of `allWagesFor`.
As part of writing this challenge, we ran right smack into one of the most
famous bugs in JavaScript land: the lost context bug. Because we've not taught
you to deal with it, we've "given" you this function. We think you can solve
the other tests with this little piece having been given to you.

If you find yourself having extra time, try researching this topic on your own.
We'll tell you all about it in our next lesson, though.

## Conclusion

This is one of the hardest topics in JavaScript. But you have hands-on
experience with the why and motivations of it! You're so much better off than
most JavaScript hackers who _never_ quite get the hang of it. It's been a lot
of growth, but this hard-won knowledge is going to help you do staggeringly
cool things. Here's a summary of execution context all in one place. Come back
whenever you need! We do!

1. Execution context is set at function call-time, implicitly or explicitly.
2. Execution context defaults to the global object unless prevented by `"use strict"`.

> 3. Execution context defaults to the containing `Object` for function expressions that are properties within that `Object*

Reviewer: ```suggestion
3. Execution context defaults to the containing `Object` for function expressions that are properties within that `Object`
```
